src/Propulsion.MessageDb/MessageDbSource.fs

```diff
@@ -0,0 +1,97 @@
+           index = reader.GetInt64(0),
+           eventType = reader.GetString(1),
+           data = (reader.GetFieldValue<byte array>(2) |> readonly),
+           meta = (reader.GetFieldValue<byte array>(3) |> readonly),
+           eventId = reader.GetGuid(4),
+           ?correlationId = readNullableString 5,
+           ?causationId = readNullableString 6,
+           timestamp = timestamp)
+
+        struct(StreamName.parse streamName, event)
+    member _.ReadCategoryMessages(category: TrancheId, fromPositionInclusive: int64, batchSize: int, ct) = task {
+        use conn = new NpgsqlConnection(connectionString)
+        do! conn.OpenAsync(ct)
+        let command = conn.CreateCommand()
+        command.CommandText <- "select
+                                   global_position, type, data, metadata, id::uuid,
+                                   (metadata::jsonb->>'$correlationId')::text,
+                                   (metadata::jsonb->>'$causationId')::text,
+                                   time, stream_name
+                                 from get_category_messages(@Category, @Position, @BatchSize);"
+        command.Parameters.AddWithValue("Category", NpgsqlDbType.Text, TrancheId.toString category) |> ignore
+        command.Parameters.AddWithValue("Position", NpgsqlDbType.Bigint, fromPositionInclusive) |> ignore
+        command.Parameters.AddWithValue("BatchSize", NpgsqlDbType.Bigint, int64 batchSize) |> ignore
+
+        let! reader = command.ExecuteReaderAsync(ct)
+        let! hasRow = reader.ReadAsync(ct)
+        let mutable hasRow = hasRow
+
+        let events = ResizeArray()
+        while hasRow do
+            events.Add(readRow reader)
+            let! nextHasRow = reader.ReadAsync(ct)
+            hasRow <- nextHasRow
+        return events.ToArray() }
+    member _.ReadCategoryLastVersion(category: TrancheId, ct) = task {
+        use conn = new NpgsqlConnection(connectionString)
+        do! conn.OpenAsync(ct)
+        let command = conn.CreateCommand()
+        command.CommandText <- "select max(global_position) from messages where category(stream_name) = @Category;"
+        command.Parameters.AddWithValue("Category", NpgsqlDbType.Text, TrancheId.toString category) |> ignore
+
+        let! reader = command.ExecuteReaderAsync(ct)
+        let! hasRow = reader.ReadAsync(ct)
+        return if hasRow then reader.GetInt64(0) else 0L
+    }
+
+module private Impl =
+    open Propulsion.Infrastructure // AwaitTaskCorrect
+
+    let readBatch batchSize (store : MessageDbCategoryReader) (category, pos) : Async<Propulsion.Feed.Core.Batch<_>> = async {
+        let! ct = Async.CancellationToken
+        let positionInclusive = Position.toInt64 pos
+        let! page = store.ReadCategoryMessages(category, positionInclusive, batchSize, ct) |> Async.AwaitTaskCorrect
+        let checkpoint = match Array.tryLast page with Some struct(_, evt) -> evt.Index | None -> positionInclusive
```

Comment:
this needs to be the global position (vs the Index, which needs to be a Stream one)
prob best for this logic to be in the ReadCategoryMessages function and for it to be passed out in a tupled result